File: tools/tour/evidence-runner.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require_relative 'evidence'
require 'tempfile'

root = File.expand_path('../..', __dir__)
inventory_file = ENV.fetch('TOUR_INVENTORY', File.join(root, 'tests/tour/inventory.tsv'))
baseline_file = ENV.fetch('TOUR_BASE_RESULTS', File.join(root, 'tests/tour/results.tsv'))
output = ENV.fetch('TOUR_EVIDENCE', File.join(root, 'tests/tour/evidence.jsonl'))
normalizer = ENV.fetch('TOUR_NORMALIZER', File.join(root, 'tools/tour/normalize.rb'))
timeout = Integer(ENV.fetch('TOUR_STEP_TIMEOUT', '30'))
bashy = ENV.fetch('BASHPP_BIN') { abort 'FATAL: BASHPP_BIN must name a clean, published bashy release binary (e.g. the output of `bashy self fetch`); a workspace dev build cannot produce evidence' }

pin = File.readlines(File.join(root, 'docs/tour/pin.tsv'), chomp: true).find { |line| !line.start_with?('#') && !line.empty? }.split("\t")
version, commit, inventory_sha = pin[1], pin[2], pin[7]
tour_root = ENV.fetch('TOUR_ROOT', File.join(`go env GOMODCACHE`.strip, "golang.org/x/website@#{version}"))
tc = File.readlines(File.join(root, 'docs/tour/toolchain.tsv'), chomp: true).find { |line| !line.start_with?('#') && !line.empty? }.split("\t")
go_bin = File.join(`GOTOOLCHAIN=#{tc[2]} go env GOROOT`.strip, 'bin/go')

inventory = File.readlines(inventory_file, chomp: true).map do |line|
  next if line.start_with?('#') || line.empty?
  f = line.split("\t")
  next unless %w[applicable_go_program build_only_go_program].include?(f[3])
  { 'path' => f[0], 'applicability' => f[3], 'bytes' => Integer(f[6]), 'sha256' => f[7] }
end.compact
abort "FATAL: executable inventory must contain exactly 97 programs" unless inventory.length == 97

baseline_rows = File.readlines(baseline_file, chomp: true).map do |line|
  next if line.start_with?('#') || line.empty?
  f = line.split("\t"); [f[0], f]
end.compact.to_h

def command_info(path, version_args: ['--version'])
  return { 'path' => path, 'present' => false, 'sha256' => nil, 'version' => nil } unless File.executable?(path)
  version, = Open3.capture2e(path, *version_args)
  { 'path' => File.realpath(path), 'present' => true, 'sha256' => TourEvidence.sha(File.binread(path)), 'version' => version.lines.first.to_s.strip }
end

go_info = command_info(go_bin, version_args: ['version']).merge('identity' => tc[3], 'pinned_sha256' => tc[4])
abort "FATAL: exact Go 1.27 binary unavailable at #{go_bin}" unless go_info['present']
abort "FATAL: Go binary is not the pinned toolchain (identity #{go_info['version']}, expected #{tc[3]})" unless go_info['version'] == tc[3]
abort "FATAL: Go binary checksum does not match docs/tour/toolchain.tsv pin" unless go_info['sha256'] == tc[4]

bashy_info_raw = command_info(bashy)
abort "FATAL: bashy binary unavailable at #{bashy}" unless bashy_info_raw['present']
identity = TourEvidence.bashy_identity(bashy_info_raw['version'])
abort "FATAL: bashy build is dirty or unpublished (#{bashy_info_raw['version'].inspect}) — evidence requires a clean published release, e.g. \`bashy self fetch\`" unless identity['published'] && !identity['dirty']
bashy_info = bashy_info_raw.merge(
  'source_revision' => identity['revision'],
  'published' => identity['published'],
  'reproducible' => identity['published'] && !identity['dirty'],
  'build_recipe' => 'bashy self fetch --version ' + identity['revision']
)

manifest = {
  'type' => 'manifest', 'schema' => TourEvidence::SCHEMA,
  'inventory' => { 'path' => 'tests/tour/inventory.tsv', 'executable_programs' => 97, 'data_sha256' => inventory_sha },
  'baseline' => { 'accepted_results' => 'tests/tour/results.tsv', 'accepted_results_sha256' => TourEvidence.sha(File.binread(baseline_file)),
                  'pin' => 'docs/tour/baseline-pin.tsv', 'pin_sha256' => TourEvidence.sha(File.binread(File.join(root, 'docs/tour/baseline-pin.tsv'))) },
  'go' => go_info,
  'bashy' => bashy_info,
  'normalizer' => { 'path' => 'tools/tour/normalize.rb', 'version' => `#{RbConfig.ruby} #{normalizer} --version`.strip,
                    'sha256' => TourEvidence.sha(File.binread(normalizer)) },
  'attempts' => 291
}

records = [manifest]
Dir.mktmpdir('tour-evidence') do |work|
  mod = File.join(work, 'module')
  helper = File.readlines(File.join(root, 'docs/tour/helpers.tsv'), chomp: true).find { |l| !l.start_with?('#') && !l.empty? }.split("\t")
  TourEvidence.materialize_module(mod, tour_root: tour_root, inventory: inventory, go_version: tc[2], helper: helper)

  inventory.each do |item|
    local = File.join(mod, item['path'])
    compiled_dir = File.join(work, 'compiled', item['path'].gsub(%r{[^\w./]}, '_'))
    FileUtils.mkdir_p(compiled_dir)
    baseline_attempt = nil
    TourEvidence::MODES.each do |mode|
      env = { 'GOTOOLCHAIN' => 'local', 'BASHY_HINTS' => 'off' }
      if mode == 'compiled'
        pipeline = TourEvidence.compiled_pipeline(bashy: bashy, go_bin: go_bin, source: local, module_dir: mod, workdir: compiled_dir, timeout: timeout, env: env)
        command, stage, raw = pipeline['command'], pipeline['stage'], pipeline['raw']
      else
        command = TourEvidence.command_for(mode, item, bashy: bashy, go_bin: go_bin, local: local)
        stage = 'run'
        raw = TourEvidence.capture(command, chdir: mod, timeout: timeout, env: env)
      end
      abort "FATAL: agent-specific Bashy advertisement entered #{item['path']} #{mode} evidence" if TourEvidence.agent_banner?(raw)
      attempt = {
        'type' => 'attempt', 'path' => item['path'], 'applicability' => item['applicability'], 'mode' => mode,
        'source' => { 'bytes' => item['bytes'], 'sha256' => item['sha256'] }, 'command' => command, 'stage' => stage,
        'spawned' => raw['spawned'], 'state' => raw['state'], 'exit' => raw['exit'],
        'raw' => { 'stdout_base64' => Base64.strict_encode64(raw['stdout']), 'stderr_base64' => Base64.strict_encode64(raw['stderr']) },
        'normalized' => TourEvidence.derived(raw, normalizer)
      }
      attempt['outcome'] = TourEvidence.outcome(attempt, mode == 'baseline' ? nil : baseline_attempt)
      if mode == 'baseline'
        accepted = baseline_rows.fetch(item['path'])
        expected_exit = item['applicability'] == 'build_only_go_program' ? Integer(accepted[6]) : Integer(accepted[7])
        build_only = item['applicability'] == 'build_only_go_program'
        expected = { 'exit' => expected_exit,
          'stdout_bytes' => build_only ? nil : Integer(accepted[8]), 'stdout_sha256' => build_only ? nil : accepted[9],
          'stderr_bytes' => build_only ? nil : Integer(accepted[10]), 'stderr_sha256' => build_only ? nil : accepted[11] }
        attempt['accepted_observation'] = expected
        authentic = attempt['exit'] == expected_exit && (build_only ? raw['stdout'].empty? && raw['stderr'].empty? :
          attempt.dig('normalized', 'stdout', 'bytes') == expected['stdout_bytes'] && attempt.dig('normalized', 'stdout', 'sha256') == expected['stdout_sha256'] &&
          attempt.dig('normalized', 'stderr', 'bytes') == expected['stderr_bytes'] && attempt.dig('normalized', 'stderr', 'sha256') == expected['stderr_sha256'])
        attempt['accepted_observation_matches'] = authentic
        baseline_attempt = attempt
      end
      records << attempt
    end
  end
end

counts = records.drop(1).group_by { |r| r['outcome'] }.transform_values(&:length)
summary = { 'type' => 'summary', 'attempts' => 291, 'programs' => 97, 'modes' => TourEvidence::MODES, 'outcomes' => counts.sort.to_h }
records << summary
root_record = { 'type' => 'root', 'algorithm' => 'sha256-canonical-jsonl', 'sha256' => TourEvidence.root(records) }
records << root_record
verdict = { 'type' => 'verdict', 'value' => counts == { 'PASS' => 291 } ? 'PASS' : 'FAIL', 'root_sha256' => root_record['sha256'] }
records << verdict
FileUtils.mkdir_p(File.dirname(output)); File.open(output, 'wb') { |io| records.each { |r| io.puts(TourEvidence.canonical(r)) } }
puts "Tour evidence #{verdict['value']}: 97 programs x 3 modes = 291 attempts; root #{root_record['sha256']}"

File: tools/tour/evidence_runner.go
// Sprint 155 / Story S155.9 / Story-ID 43af37063b09.
//
// Produces tests/tour/evidence.jsonl (tour-evidence/v2) — the port of
// tools/tour/evidence-runner.rb. Superseded by the three-mode executor; kept
// because its retained ledger is historical failure evidence and its
// validator is harness-wired.
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

func evidenceCommandInfo(path string, versionArgs ...string) map[string]any {
	if !isExecutable(path) {
		return map[string]any{"path": path, "present": false, "sha256": nil, "version": nil}
	}
	out, _ := combinedOutput(path, versionArgs...)
	real := path
	if r, err := filepath.EvalSymlinks(path); err == nil {
		real, _ = filepath.Abs(r)
	}
	return map[string]any{"path": real, "present": true, "sha256": shaFile(path), "version": strings.TrimSpace(firstLine(out))}
}

func cmdEvidenceRunner(root string) int {
	inventoryFile := envOr("TOUR_INVENTORY", filepath.Join(root, "tests/tour/inventory.tsv"))
	baselineFile := envOr("TOUR_BASE_RESULTS", filepath.Join(root, "tests/tour/results.tsv"))
	output := envOr("TOUR_EVIDENCE", filepath.Join(root, "tests/tour/evidence.jsonl"))
	normalizer := envOr("TOUR_NORMALIZER", filepath.Join(root, normalizerPath))
	timeout := float64(mustInt(envOr("TOUR_STEP_TIMEOUT", "30")))
	bashy, ok := os.LookupEnv("BASHPP_BIN")
	if !ok {
		return abortf("FATAL: BASHPP_BIN must name a clean, published bashy release binary (e.g. the output of `bashy self fetch`); a workspace dev build cannot produce evidence")
	}

	pin := firstDataRow(filepath.Join(root, "docs/tour/pin.tsv"))
	version, inventorySHA := pin[1], pin[7]
	tourRoot := envOr("TOUR_ROOT", filepath.Join(shellOutput(nil, "go", "env", "GOMODCACHE"), "golang.org/x/website@"+version))
	tc := firstDataRow(filepath.Join(root, "docs/tour/toolchain.tsv"))
	goBin := filepath.Join(shellOutput(map[string]string{"GOTOOLCHAIN": tc[2]}, "go", "env", "GOROOT"), "bin/go")

	inventory, _ := evidenceInventory(inventoryFile)
	if len(inventory) != 97 {
		return abortf("FATAL: executable inventory must contain exactly 97 programs")
	}
	baselineRows := map[string][]string{}
	for _, f := range tsvRowsLoose(baselineFile) {
		baselineRows[field(f, 0)] = f
	}

	goInfo := evidenceCommandInfo(goBin, "version")
	goInfo["identity"] = tc[3]
	goInfo["pinned_sha256"] = tc[4]
	if !truthy(goInfo["present"]) {
		return abortf("FATAL: exact Go 1.27 binary unavailable at %s", goBin)
	}
	if goInfo["version"] != tc[3] {
		return abortf("FATAL: Go binary is not the pinned toolchain (identity %s, expected %s)", toS(goInfo["version"]), tc[3])
	}
	if goInfo["sha256"] != tc[4] {
		return abortf("FATAL: Go binary checksum does not match docs/tour/toolchain.tsv pin")
	}

	bashyRaw := evidenceCommandInfo(bashy, "--version")
	if !truthy(bashyRaw["present"]) {
		return abortf("FATAL: bashy binary unavailable at %s", bashy)
	}
	identity := bashyIdentity(asString(bashyRaw["version"]))
	if !(truthy(identity["published"]) && !truthy(identity["dirty"])) {
		return abortf("FATAL: bashy build is dirty or unpublished (%s) — evidence requires a clean published release, e.g. `bashy self fetch`", inspectString(asString(bashyRaw["version"])))
	}
	bashyInfo := map[string]any{}
	for k, v := range bashyRaw {
		bashyInfo[k] = v
	}
	bashyInfo["source_revision"] = identity["revision"]
	bashyInfo["published"] = identity["published"]
	bashyInfo["reproducible"] = truthy(identity["published"]) && !truthy(identity["dirty"])
	bashyInfo["build_recipe"] = "bashy self fetch --version " + toS(identity["revision"])

	normalizerRel, _ := filepath.Rel(root, normalizer)
	manifest := map[string]any{
		"type": "manifest", "schema": evidenceSchema,
		"inventory": map[string]any{"path": "tests/tour/inventory.tsv", "executable_programs": int64(97), "data_sha256": inventorySHA},
		"baseline": map[string]any{"accepted_results": "tests/tour/results.tsv", "accepted_results_sha256": shaFile(baselineFile),
			"pin": "docs/tour/baseline-pin.tsv", "pin_sha256": shaFile(filepath.Join(root, "docs/tour/baseline-pin.tsv"))},
		"go":    goInfo,
		"bashy": bashyInfo,
		"normalizer": map[string]any{"path": filepath.ToSlash(normalizerRel), "version": normalizerVersion,
			"sha256": shaFile(normalizer)},
		"attempts": int64(291),
	}

	records := []map[string]any{manifest}
	work, err := os.MkdirTemp("", "tour-evidence")
	if err != nil {
		return abortf("FATAL: %v", err)
	}
	defer os.RemoveAll(work)
	mod := filepath.Join(work, "module")
	helper := firstDataRow(filepath.Join(root, "docs/tour/helpers.tsv"))
	if err := evidenceMaterializeModule(mod, tourRoot, inventory, tc[2], helper); err != nil {
		return abortf("FATAL: %v", err)
	}

	for _, item := range inventory {
		local := filepath.Join(mod, item.Path)
		compiledDir := filepath.Join(work, "compiled", unsafePathCharRE.ReplaceAllString(item.Path, "_"))
		os.MkdirAll(compiledDir, 0o755)
		var baselineAttempt map[string]any
		for _, mode := range modes {
			env := map[string]string{"GOTOOLCHAIN": "local", "BASHY_HINTS": "off"}
			var command any
			var stage string
			var raw EvidenceCapture
			if mode == "compiled" {
				pipeline := evidenceCompiledPipeline(bashy, goBin, local, mod, compiledDir, timeout, env)
				command, stage, raw = pipeline.Command, pipeline.Stage, pipeline.Raw
			} else {
				argv := evidenceCommandFor(mode, item.Applicability, bashy, goBin, local)
				command, stage = anyList(argv), "run"
				raw = evidenceCapture(argv, mod, timeout, env)
			}
			if agentBanner(raw.Stderr) {
				return abortf("FATAL: agent-specific Bashy advertisement entered %s %s evidence", item.Path, mode)
			}
			attempt := map[string]any{
				"type": "attempt", "path": item.Path, "applicability": item.Applicability, "mode": mode,
				"source": map[string]any{"bytes": item.Bytes, "sha256": item.SHA256}, "command": command, "stage": stage,
				"spawned": raw.Spawned, "state": raw.State, "exit": raw.Exit,
				"raw":        map[string]any{"stdout_base64": b64(raw.Stdout), "stderr_base64": b64(raw.Stderr)},
				"normalized": evidenceDerived(raw.Stdout, raw.Stderr),
			}
			var base map[string]any
			if mode != "baseline" {
				base = baselineAttempt
			}
			attempt["outcome"] = evidenceOutcome(attempt, base)
			if mode == "baseline" {
				accepted := baselineRows[item.Path]
				buildOnly := item.Applicability == "build_only_go_program"
				expected := expectedAccepted(accepted, buildOnly)
				attempt["accepted_observation"] = expected
				authentic := jsonEqual(attempt["exit"], expected["exit"])
				if buildOnly {
					authentic = authentic && len(raw.Stdout) == 0 && len(raw.Stderr) == 0
				} else {
					authentic = authentic && jsonEqual(dig(attempt, "normalized", "stdout", "bytes"), expected["stdout_bytes"]) &&
						jsonEqual(dig(attempt, "normalized", "stdout", "sha256"), expected["stdout_sha256"]) &&
						jsonEqual(dig(attempt, "normalized", "stderr", "bytes"), expected["stderr_bytes"]) &&
						jsonEqual(dig(attempt, "normalized", "stderr", "sha256"), expected["stderr_sha256"])
				}
				attempt["accepted_observation_matches"] = authentic
				baselineAttempt = attempt
			}
			records = append(records, attempt)
		}
	}

	counts := countBy(records[1:], "outcome")
	summary := map[string]any{"type": "summary", "attempts": int64(291), "programs": int64(97), "modes": anyList(modes), "outcomes": counts}
	records = append(records, summary)
	rootRecord := map[string]any{"type": "root", "algorithm": "sha256-canonical-jsonl", "sha256": ledgerRoot(records)}
	records = append(records, rootRecord)
	value := "FAIL"
	if jsonEqual(counts, map[string]any{"PASS": int64(291)}) {
		value = "PASS"
	}
	records = append(records, map[string]any{"type": "verdict", "value": value, "root_sha256": rootRecord["sha256"]})
	if err := writeLedger(output, records); err != nil {
		return abortf("FATAL: %v", err)
	}
	fmt.Printf("Tour evidence %s: 97 programs x 3 modes = 291 attempts; root %s\n", value, rootRecord["sha256"])
	return 0
}

// expectedAccepted renders the accepted-observation record from a
// tests/tour/results.tsv row.
func expectedAccepted(row []string, buildOnly bool) map[string]any {
	exitField := 7
	if buildOnly {
		exitField = 6
	}
	expected := map[string]any{"exit": mustInt(field(row, exitField)),
		"stdout_bytes": nil, "stdout_sha256": nil, "stderr_bytes": nil, "stderr_sha256": nil}
	if !buildOnly {
		expected["stdout_bytes"] = mustInt(field(row, 8))
		expected["stdout_sha256"] = field(row, 9)
		expected["stderr_bytes"] = mustInt(field(row, 10))
		expected["stderr_sha256"] = field(row, 11)
	}
	return expected
}
